# Analyzing the Trend of the Used Car Deals – Exploring eBay Car Sales Data

## Introduction

[_eBay Kleinanzeigen_](https://www.ebay-kleinanzeigen.de/) is a classifieds section of the German eBay website. In this project, we will use the used cars dataset from _eBay Kleinanzeigen_ to analyze the used cars ads created in Germany between 11th June 2015 and 7th April 2016.

The original dataset uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka) is no longer available on Kaggle, but it is still accessible on [data.world](https://data.world/data-society/used-cars-data).

In this project, we use the _modified_ version of the original dataset, which was prepared by Dataquest. The modifications are as follows:
- _50,000 data points_ have been sampled from the original full dataset, which originally contained 370,000 data points.
- The dataset _was dirtied_ slightly — to make it resembles a scraped dataset before data cleaning (the original dataset on Kaggle had been cleaned).

The explanations for the columns of the `autos.csv` dataset are as follows:
- `dateCrawled`: The date when this ad was first crawled
- `name`: Name of the car
- `seller`: A private seller or a dealer
- `offerType`: The listing type
- `price`: The asking price for selling the car
- `abtest`: Whether the listing is included in an A/B test
- `vehicleType`: The vehicle type
- `yearOfRegistration`: The year in which the car was first registered
- `gearbox`: The transmission type
- `powerPS`: The power of the car in PS
- `model`: The car model name
- `odometer`: How many kilometres the car has driven
- `monthOfRegistration`: The month in which the car was first registered
- `fuelType`: What type of fuel the car uses
- `brand`: The brand of the car
- `notRepairedDamage`: Whether the car has damage which still needs to be repaired
- `dateCreated`: The eBay listing creation date
- `nrOfPictures`: The number of pictures in the ad
- `postalCode`: The postal code for the location of the vehicle
- `lastSeenOnline`: When the crawler saw this ad last online

The content of the dataset is in German, as the data was originally scraped from the German eBay website.

### The Goal of the Project
The goal of this project is to clean the data and analyze the trend of the used car deals.

### Summary of Results
We cleaned the dataset and observed that March and beginning of April 2016 are the busiest time for ad creation. We also found a lack of correlation between mean price and mean mileage by brand. The distribution of `date_crawled` and `last_seen` are fairly consistent throughout the whole data sampling period.

## Reading the Data and Cleaning the Column Names
We read the `autos.csv` file into *pandas* and assign it to the variable name `autos`. The file could read by using `encoding=Latin-1` instead of the default encoding `UTF-8`.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

Let's briefly examine the data:

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Here are what we observe:
- The `autos` dataset contains *50,000 rows and 20 columns*.
- Most of the entries are *string objects (15)*, whereas only several are *integers (5)*.
- Some columns have *null values (5)* but in low percentage.
- The column names use *camelcase* instead of *snakecase*, which is the preferred case for Python.

We will transform the column names from camelcase to the preferred snakecase and modify some of the column names to make them more descriptive.

The original column names are as follows:

In [5]:
# Show the original column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We copy the array of the column names and edit it, and eventually assign the modified column names back to the `autos.columns`:

In [6]:
# Edit the column names
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

To verify whether the column names have been modified, we call `autos.head()` to examine the column names of the `autos` dataframe. Alternatively, we can also verify the column names by running `autos.columns`.

In [7]:
# Display the data of the first 5 rows
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The output shows that the column names have been converted to snakecase and several names have been reworded.

## Initial Exploration and Cleaning

Now, we will explore the data a little bit more to determine the necessary basic data cleaning tasks. We use `autos.describe()` with `include='all`' to examine the descriptive statistics for both categorical and numeric columns.

In [8]:
# Display the descriptive statistics
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Cleaning the Non-Value-Added Columns

Based on the output, we notice that the `nr_of_pictures` column has only one value, which is `0`.

In [9]:
# Show the unique value
autos['nr_of_pictures'].unique()

array([0])

Since a single-value column would not add additional value to our data analysis, we will remove `nr_of_pictures` from our dataset.

In [10]:
# Remove `nr_of_pictures` from the dataset
autos.drop(columns=['nr_of_pictures'], inplace=True)

To verify whether the `nr_of_pictures` column has been removed, we did the following:
- Run `autos.shape`: The number of columns has been reduced from 20 to *19*.
- Run `autos.columns`: The `nr_of_pictures` column has disappeared.
 
These results confirm that the `nr_of_pictures` column has been removed from the dataset.

In [11]:
# Show the number of rows and columns
autos.shape

(50000, 19)

In [12]:
# Show the column names
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

We also observe that only one out of the 50,000 entries for the `seller` column is `gewerblich` (or `commercial` in English). The remaining are `privat` (or `private` in English). We can delete the entry for the `gewerblich` to focus on the entries of the `privat` dealers.

Same as the observation for the `seller`, the `offer_type` column shows 49,999 entries for `Angebot` (or `Offer` in English), while only one entry for `Gesuch` (or `Request` in English). Since we focus on the `Angebot` (`Offer`), we will delete the entry for the `Gesuch` to prevent it from messing up our data analysis.

First, we isolate the rows for the `gewerblich` (`commercial`) seller and `Gesuch` (`Request`) offer_type:

In [13]:
# Isolate `commercial` seller
seller_commercial = autos[autos['seller'] == 'gewerblich']

seller_commercial

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price   abtest vehicle_type  registration_year  \
7738  gewerblich    Angebot  $100  control        kombi               2000   

      gearbox  power_ps   model   odometer  registration_month fuel_type  \
7738  manuell         0  megane  150,000km                   8    benzin   

        brand unrepaired_damage           ad_created  postal_code  \
7738  renault               NaN  2016-03-15 00:00:00        65232   

                last_seen  
7738  2016-04-06 17:15:37

In [14]:
# Isolate `Request` offer_type
offer_type_request = autos[autos['offer_type'] == 'Gesuch']

offer_type_request

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      abtest vehicle_type  registration_year gearbox  power_ps        model  \
17541   test          bus               2005     NaN         0  transporter   

        odometer  registration_month fuel_type       brand unrepaired_damage  \
17541  150,000km                   0       NaN  volkswagen               NaN   

                ad_created  postal_code            last_seen  
17541  2016-04-03 00:00:00        29690  2016-04-05 15:16:06

Next, we remove these two rows by using `autos.drop` and the indexes of these rows:

In [15]:
# Remove the rows of `commercial` seller and `Request` offer_type from our dataset
autos.drop(index=[7738, 17541], inplace=True)

The result from `autos.shape` shows that the number of rows has been reduced from 50,000 to 49,998. This indicates that two rows have been deleted from the dataset.

In [16]:
# Show the number of rows and columns
autos.shape

(49998, 19)

To examine whether these two deleted rows are the rows that we wanted to delete, we call these rows by using their index labels. The results show that the indexes have been updated — it confirms that the two rows that we do not intend to keep have been deleted.

In [17]:
# Index 7738 was the original index for `commercial` seller row
autos.iloc[7738]

date_crawled                                        2016-03-17 14:47:29
name                  ***_SMART_forTwo_cabrio_softouch_passion___sup...
seller                                                           privat
offer_type                                                      Angebot
price                                                           $10,999
abtest                                                          control
vehicle_type                                                     cabrio
registration_year                                                  2014
gearbox                                                       automatik
power_ps                                                              0
model                                                            fortwo
odometer                                                       20,000km
registration_month                                                    2
fuel_type                                                       

In [18]:
# Index 17541 was the original index for Request` offer_type row
autos.iloc[17541]

date_crawled                            2016-03-11 16:45:06
name                  Seat_Ibiza_1.9_TDI_Sport_SELTEN_!_TÜV
seller                                               privat
offer_type                                          Angebot
price                                                $1,950
abtest                                                 test
vehicle_type                                     kleinwagen
registration_year                                      2002
gearbox                                             manuell
power_ps                                                110
model                                                 ibiza
odometer                                          150,000km
registration_month                                        1
fuel_type                                            diesel
brand                                                  seat
unrepaired_damage                                      nein
ad_created                              

### Converting Numeric Data Stored as String Object to Integer Datatype

We notice that this numeric data stored as string objects, and we need to convert them to integers for analysis purpose:
- `price`
- `odometer`

In [19]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49998 entries, 0 to 49999
Data columns (total 19 columns):
date_crawled          49998 non-null object
name                  49998 non-null object
seller                49998 non-null object
offer_type            49998 non-null object
price                 49998 non-null object
abtest                49998 non-null object
vehicle_type          44903 non-null object
registration_year     49998 non-null int64
gearbox               47319 non-null object
power_ps              49998 non-null int64
model                 47240 non-null object
odometer              49998 non-null object
registration_month    49998 non-null int64
fuel_type             45517 non-null object
brand                 49998 non-null object
unrepaired_damage     40171 non-null object
ad_created            49998 non-null object
postal_code           49998 non-null int64
last_seen             49998 non-null object
dtypes: int64(4), object(15)
memory usage: 7.6+ MB


We build `string_to_integer` function to convert columns with the numeric data stored as *string* to *integer* datatype. It also removes unit and `,` (thousands separator ) from the numeric data.

In [20]:
# `column` is the name of the column; `unit` is the non-digit character to be removed from the value
def string_to_integer(column, unit):
    autos[column] = autos[column].str.replace(unit, '').str.replace(',', '').astype(int)
    return autos[column]

Next, we use `string_to_integer` to convert `price` and `odometer` to integer and verify the conversion by printing their datatypes.

In [21]:
# Convert the `price` from string object to integer
string_to_integer('price', '$')

# Convert the `odometer` from string object to integer
string_to_integer('odometer', 'km')

# Check their dtypes after conversion
# Alternatively, we can also use `autos.info()`
print('The dtype for:')
print('    - price:      ', autos['price'].dtype)
print('    - odometer:   ', autos['odometer'].dtype)

The dtype for:
    - price:       int64
    - odometer:    int64


We also update the column names by including their corresponding units — an important piece of information to keep.

In [22]:
# Rename the columns for `price` and `odometer`
autos.rename({'price': 'price_$', 'odometer' : 'odometer_km'}, axis=1, inplace=True)

Next, we confirm that all the changes have been made on the dataset by examining `autos.csv`.

In [23]:
# Display the data
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

## Exploring the Price Column

After the initial data cleaning, we explore the data in detailed — specifically examine for data that is illogical. We start by analyzing the `price_$` column.

In [24]:
# Display the descriptive statistics
autos['price_$'].describe()

count    4.999800e+04
mean     9.840435e+03
std      4.811140e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

The statistic information above is displayed in scientific notation, which is not so friendly to read in the price context. Therefore, we will change the display format to floats with one decimal place and include a thousand separator (`,`) by using `pd.options.display.float_format` method.

In [25]:
# Change the format to show floats with 1 decimal place and include a thousand separator
pd.options.display.float_format = '{:,.1f}'.format

# Use the following to reset the format
# pd.reset_option('^display.', silent=True)

After adjusting the format, we reprint the descriptive statistics:

In [26]:
# Display the descriptive statistics with updated format
autos['price_$'].describe()

count       49,998.0
mean         9,840.4
std        481,114.0
min              0.0
25%          1,100.0
50%          2,950.0
75%          7,200.0
max     99,999,999.0
Name: price_$, dtype: float64

In [27]:
print('The number of unique values:   ', autos['price_$'].unique().shape[0])

The number of unique values:    2357


### Outliers in the Minumum Value

We notice that *0* for the min price of listed cars is unreasonable, as it is basically *FREE*!

Surprisingly, the frequency for 0 is very high — 1420 entries — therefore, they do not seem to be entry errors.

In [28]:
# Examine the value counts of `price_$` and sort their index in a descending manner
autos['price_$'].value_counts().sort_index().head(10)

0     1420
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price_$, dtype: int64

We extract the details of these entries and give it a variable name `min_price` for further investigation.

In [29]:
min_price = autos[autos['price_$'] == 0]

min_price

date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
118    2016-03-12 05:03:00  VW_Sharan_V6_204_PS_Karosse_Rohkarosse_mit_Pap...   
146    2016-03-22 23:59:28                                    Ford_Fiesta_rot   
167    2016-04-02 19:43:45  Suche_VW_Multivan_Innenausstattung_Set_oder_TE...   
180    2016-03-19 10:50:25                                       Zu_verkaufen   
226    2016-03-25 23:52:12                      Porsche_911_S_Targa__67er_SWB   
234    2016-03-05 16:56:22                                   Fiat_Punto_Sport   
248    2016-03-08 22:57:53                             VW_Passat_zu_verkaufen   
259    2016-04-03 23:49:58                      guenstiges_Auto_/_auch_defekt   
301    2016-03-08 20:37:59                         Kaufe_alle_Autos_bietet_an   
323    2016-03-15 16:54:59                                W220_Unfallfahrzeug   
327    2016-04-03 13:49:21              BMW_Cabrio_318i_/_BMW_318_tds_Touring   
347    2016-04-02 11:48:02                                  Volvo_TP21__Sugga   
366    2016-03-21 20:36:25  Ford_Escort_Express_75__LKW_Zulussung___Diesel...   
389    2016-03-08 11:38:52                                    Vw_polo_classic   
418    2016-03-29 14:43:24         Fiat_SCUDO_8_Sitzer_Bus__Diesel_JTD__80_KW   
427    2016-03-12 11:44:47                 TOP_ZUSTAND_GrunePlakette_Euro4!!!   
430    2016-03-18 23:52:40                                  Winterraeder_FORD   
449    2016-04-03 20:59:59                           Tausche_gegen_ein_smart.   
487    2016-03-14 21:41:36  MAG_MICH_KEINER_HABEN_Opel_Omega_B_Caravan_2.0...   
502    2016-03-25 23:55:24                               C32_AMG_C55_C43_w203   
519    2016-03-28 08:53:58  Verkaufe/Tausche_Seat_Inca__VW_Caddy__aus_2002...   
565    2016-03-20 20:49:36     Tausche_Audi_S3_8p_gegen_Golf_VI_Gti_mit_xenon   
593    2016-04-01 21:44:15                              CLK_200__BRABUS_Optik   
660    2016-03-29 21:44:40                                       Opel_Corsa_B   
661    2016-03-18 22:49:48               Renault_Megane_Scenic_1.6_Tuev_12/17   
...                    ...                                                ...   
48882  2016-03-14 23:58:47                                         SUCHE_AUTO   
48889  2016-04-01 10:50:52                           Audi_A4_____TueV_09.2016   
48891  2016-03-20 22:50:14        MB_E_220_CDI__TAUSCHEN_nur_SUV/GELÄNDEWAGEN   
48982  2016-03-08 12:48:33                           Bmw_525D_Sport_Pro..Navi   
48985  2016-03-27 08:55:25                             Stockcar_Autos_Peugeot   
49043  2016-03-07 11:48:55                              Opel_Vectra_B_Caravan   
49082  2016-03-13 17:06:58         Passat_35i_b4_in_candy_weiss_original_lack   
49254  2016-03-21 10:49:57                           Corrado_G60_mit_2e_Motor   
49301  2016-03-20 20:52:40                           Golf_3_Variant_Kombi_1.8   
49342  2016-03-09 20:47:01  Audi_80_b4_S_Color_und_BMW_318ti_compact_Tausc...   
49350  2016-03-19 23:57:15                                     Golf_II_Diesel   
49404  2016-04-04 03:03:16                                Golf_2_BBS_LEDER_KW   
49408  2016-03-30 20:57:06                                         OPEL_ASTRA   
49479  2016-04-01 23:55:01                            Tausch_mit_frischen_TÜV   
49496  2016-03-26 13:55:28                                        Bmw_e39_520   
49504  2016-03-13 19:42:57                                 Meriva_Lpg_1_6_16v   
49507  2016-03-16 17:38:32                 BMW_528i_E39_Schaltgetriebe_Unfall   
49525  2016-03-09 17:59:02              Bmw_e

When we analyze the number of counts of the `registration_year` of the cars under `min_price`, we found that some of the cars are very new, for example, *96 of them were registered in 2016*. Thus, it is unreasonable to us that the asking price is only 0.

Interestingly, there are also *60 cars registered in 2017*, which is illogical, as the data was crawled in 2016. Because of that, we need to delete the entries with registration after 2016.

In [30]:
min_price['registration_year'].value_counts().head(10)

2000    198
1999    101
2016     96
1998     90
1995     85
2005     78
1997     77
1996     71
2001     66
2017     60
Name: registration_year, dtype: int64

To investigate whether those *'free'* cars belong to cheaper or unknown brands, we analyze their brands.

In [31]:
min_price['brand'].value_counts().head(10)

volkswagen        347
opel              183
bmw               154
audi              115
ford               95
mercedes_benz      81
renault            79
sonstige_autos     72
fiat               45
peugeot            26
Name: brand, dtype: int64

Given that the top 10 brands of the *'free'* cars include many luxurious brands (e.g. BMW, Audi and Mercedes-Benz), as well as decent/well-known brands (e.g. Volkswagen, Ford, Renault, etc.), it is bizarre that their asking price is only 0.

eBay is an auction site, which means that in theory, a seller could propose the lowest asking price at 0. However, based on the several points that we just mentioned, **0 is still an unrealistically low asking price, especially for new (or not too old) cars and decent brands**. Therefore, we will **remove their entries** from our dataset.

### Outliers in the Maximum Value

In [32]:
autos['price_$'].describe()

count       49,998.0
mean         9,840.4
std        481,114.0
min              0.0
25%          1,100.0
50%          2,950.0
75%          7,200.0
max     99,999,999.0
Name: price_$, dtype: float64

Based on the descriptive statistics for `price_$`, the maximum value is *99,999,999*, which far exceeds the 7,200 of the third quartile (Q3 or 75% of the data). Considering that this is a very old car registered in 1999, the asking price is too high, and it sounds illogical. Therefore, this entry is definitely an outlier that needs to be removed.

In [33]:
autos[autos['price_$'] == 99999999]

date_crawled                         name  seller offer_type  \
39705  2016-03-22 14:58:27  Tausch_gegen_gleichwertiges  privat    Angebot   

        price_$   abtest vehicle_type  registration_year    gearbox  power_ps  \
39705  99999999  control    limousine               1999  automatik       224   

          model  odometer_km  registration_month fuel_type          brand  \
39705  s_klasse       150000                   9    benzin  mercedes_benz   

      unrepaired_damage           ad_created  postal_code            last_seen  
39705               NaN  2016-03-22 00:00:00        73525  2016-04-06 05:15:30

To find out more about other potential outliers at the upper end of the price, we investigate the data of the top 20 most expensive cars.

In [34]:
autos['price_$'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_$, dtype: int64

In [35]:
# Based on the sorting result, the 20th most expensive car costs $197,000
# Call for the details of the entries of the top 20 most expensive cars
max_price = autos[autos['price_$'] >= 197000].sort_values('price_$', ascending=False)

max_price

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
36818  2016-03-27 18:37:37                                        Porsche_991   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   
38299  2016-03-28 22:25:25                                Glas_BMW_mit_Wasser   
37840  2016-03-21 10:50:12                                        Porsche_997   
40918  2016-03-20 18:40:05                             Porsche_911_991_GT3_RS   
43668  2016-03-16 18:47:26  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   

       seller offer_type   price_$   abtest vehicle_type  registration_year  \
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
47598  privat    Angebot  12345678  control    limousine               2001   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
2897   privat    Angebot  11111111     test    limousine               1973   
24384  privat    Angebot  11111111     test          NaN               1995   
11137  privat    Angebot  10000000  control        coupe               1960   
47634  privat    Angebot   3890000     test        coupe               2006   
7814   privat    Angebot   1300000  control        coupe               1992   
22947  privat    Angebot   1234566  control        kombi               1999   
43049  privat    Angebot    999999     test          bus               1981   
514    privat    Angebot    999999     test        kombi               2009   
37585  privat    Angebot    999990     test    limousine               1985   
36818  privat    Angebot    350000  control        coupe               2016   
14715  privat    Angebot    345000  control       cabrio               2012   
34723  privat    Angebot    299000     test        coupe               1977   
35923  privat    Angebot    295000     test       cabrio               2015   
12682  privat    Angebot    265000  control        coupe               2016   
47337  privat    Angebot    259000     test       cabrio               2001   
38299  privat    Angebot    250000     test          NaN               2015   
37840  privat    Angebot    220000     test        coupe               2008   
40918  privat    Angebot    198000     test        coupe               2015   
43668  privat    Angebot    197000  control        coupe               1998   

         gea

Based on the output, we noticed that there is a huge price gap between USD 350,000 (Porsche 991, `index: 36818`) and 999,990 (Volkswagen Jetta GT, `index: 37585`). Most of the cars equal to or above USD 999,990 are fairly old cars (registered between the year 1960 to 2009), with several exceptions:
- 2014, Citroen C4, `index: 42221` — USD 27,322,222
- 2017, Fiat Punto, `index: 27371` — USD 12,345,678
- 2018, Volvo v40, `index: 39377` — USD 12,345,678

The asking price for Citroen C4 on eBay does not make sense, as its market price in 2016 was only between USD 9,314 to 12,454 (GBP 6,899 to 9,225) (Ref: [Citroen C4 2016](https://priceanycar.com/Citroen/C4/2016/)). The car entries for years 2017 and 2018 are questionable, as the crawling date was 2016 and the prices (12,345,678) seem to be invented number.

Meanwhile, a quick search on the internet browser shows that the price of the most expensive cars in 2016 are:
- USD 2.5M to USD 4.8M (Ref: [Groco](https://groco.com/article/top-8-most-expensive-cars-in-2016/))
- USD 3.4M to USD 38M (Ref: [GlobalCarsBrands](https://www.globalcarsbrands.com/top-10-most-expensive-cars-in-the-world/))

If a brand new luxurious car in 2016 costs above one million, any used cars cost close to or above 1 million in the dataset are clearly unrealistically high. Therefore, we decided to keep the entries with the price between USD 1 and USD 350,000, i.e. exclude entries at USD 0 and above USD 350,000.

We are aware of the fact that a better filter may involve more manual checking regarding whether the asking price is realistic — similar to what we just did to the data above USD 999,990. However, it is not so realistic to do so, as there are thousands of entries in the dataset. Therefore, even though our filter is still not perfect, yet it is fairly effective.

In [36]:
# Keep the rows with prices between 1 and 350,000, while discarding the rows with other values
autos = autos[autos['price_$'].between(1,350000)]

autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

In [37]:
# Display the descriptive statistics
autos.describe()

price_$  registration_year  power_ps  odometer_km  registration_month  \
count  48,564.0           48,564.0  48,564.0     48,564.0            48,564.0   
mean    5,889.1            2,004.8     117.2    125,769.6                 5.8   
std     9,059.9               88.6     200.7     39,788.9                 3.7   
min         1.0            1,000.0       0.0      5,000.0                 0.0   
25%     1,200.0            1,999.0      71.0    125,000.0                 3.0   
50%     3,000.0            2,004.0     107.0    150,000.0                 6.0   
75%     7,490.0            2,008.0     150.0    150,000.0                 9.0   
max   350,000.0            9,999.0  17,700.0    150,000.0                12.0   

       postal_code  
count     48,564.0  
mean      50,975.5  
std       25,747.2  
min        1,067.0  
25%       30,657.0  
50%       49,716.0  
75%       71,665.0  
max       99,998.0

In [38]:
# Display the lowest price in an ascending manner
autos['price_$'].value_counts().sort_index().head()

1    156
2      3
3      1
5      2
8      1
Name: price_$, dtype: int64

In [39]:
# Display the highest price in an descending manner
autos['price_$'].value_counts().sort_index(ascending=False).head()

350000    1
345000    1
299000    1
295000    1
265000    1
Name: price_$, dtype: int64

Based on the cells above, we confirm that the entries at price 0 and above 350,000 have been removed. 

## Exploring the Odometer Column

Next, we analyze the `odometer_km` column.

In [40]:
autos['odometer_km'].describe()

count    48,564.0
mean    125,769.6
std      39,788.9
min       5,000.0
25%     125,000.0
50%     150,000.0
75%     150,000.0
max     150,000.0
Name: odometer_km, dtype: float64

In [41]:
print('The number of unique values:   ', autos['odometer_km'].unique().shape[0])

The number of unique values:    13


In [42]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31413
Name: odometer_km, dtype: int64

There are 13 unique values in `odometer_km` column. 

The `value_counts` result shows that there is no outlier. The number of counts skews towards the higher end with the highest counts being the maximum value, 150,000 km. This is followed by the second and third highest counts being 125,000 and 100,000 km, which are the second and third maximum values, respectively. 

Overall, the data looks good.   

## Exploring the Date Columns

Now, we will explore the date columns, which are:
- `date_crawled`
- `ad_created`
- `last_seen`
- `registration_year`
- `registration_month`

Let's look at the date formats of the three date columns:

In [43]:
# Display the first five rows of the three date columns
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The output shows that the values in date columns are displayed in `%Y-%m-%d %H:%M%S` format, where the initial 10 characters represent the date (e.g. `2016-03-26`).

To investigate the distribution of the date values in these three columns as percentages, we chain the methods as such:
- `str[:10]`: To extract the date
- `value_counts(normalize=True, dropna=False)`: To get the relative frequency of the unique values and include the missing values
- `sort_index()`: To sort by date in ascending order (default)
- `*100`: To convert the relative frequency to the percentage

In the earlier cells, we changed the output format of the descriptive statistics from scientific notation to floats with 1 decimal place and included a thousand separator. Now, we revert the format setting to make the analysis of the value in percentages easier.

In [44]:
# Reset the float format setting
pd.reset_option('^display.', silent=True)

In [45]:
# 'date_crawled' — To get the distribution of date values as percentages
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100

2016-03-05    2.532740
2016-03-06    1.404332
2016-03-07    3.601433
2016-03-08    3.329627
2016-03-09    3.309035
2016-03-10    3.218433
2016-03-11    3.257557
2016-03-12    3.692035
2016-03-13    1.567004
2016-03-14    3.654971
2016-03-15    3.426406
2016-03-16    2.961041
2016-03-17    3.162837
2016-03-18    1.291080
2016-03-19    3.477885
2016-03-20    3.788815
2016-03-21    3.737336
2016-03-22    3.298740
2016-03-23    3.222552
2016-03-24    2.934272
2016-03-25    3.160778
2016-03-26    3.220493
2016-03-27    3.109299
2016-03-28    3.486121
2016-03-29    3.409933
2016-03-30    3.368751
2016-03-31    3.183428
2016-04-01    3.368751
2016-04-02    3.547896
2016-04-03    3.860885
2016-04-04    3.648793
2016-04-05    1.309612
2016-04-06    0.317107
2016-04-07    0.140021
Name: date_crawled, dtype: float64

The distribution of `date_crawled` is quite consistent throughout the whole crawling period — **it fluctuates within 3% most of the time** — with several exceptions. The lower `date_crawled` percentage from 2016-04-05 to 2016-04-07 correlates with the lower `ad_created` percentage (see the next cell). This indicates that the percentage of `ad_created` affect the percentage of `date_crawled`. Overall, the crawling efficiency seems to be fairly good.

In [46]:
# 'ad_created' — To get the distribution of date values as percentages
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()*100

2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
2015-12-30    0.002059
2016-01-03    0.002059
2016-01-07    0.002059
2016-01-10    0.004118
2016-01-13    0.002059
2016-01-14    0.002059
2016-01-16    0.002059
2016-01-22    0.002059
2016-01-27    0.006177
2016-01-29    0.002059
2016-02-01    0.002059
2016-02-02    0.004118
2016-02-05    0.004118
2016-02-07    0.002059
2016-02-08    0.002059
2016-02-09    0.002059
2016-02-11    0.002059
2016-02-12    0.004118
2016-02-14    0.004118
2016-02-16    0.002059
2016-02-17    0.002059
2016-02-18    0.004118
2016-02-19    0.006177
2016-02-20    0.004118
2016-02-21    0.006177
                ...   
2016-03-09    3.315213
2016-03-10    3.189605
2016-03-11    3.290503
2016-03-12    3.675562
2016-03-13    1.700848
2016-03-14    3.519068
2016-03-15    3.399638
2016-03-16    3.012520
2016-03-17    3.127831
2016-03-18    1.359031
2016-03-19    3.368751
2016-03-20    3.794992
2016-03-21 

March 2016 and beginning of April 2016 were the busiest month for used cars ads creation, with around 3.+ % of daily new ads.  The percentages of daily ads creation decreased drastically from 2016-04-05 to 2016-04-07 from 1.181945% to 0.125607%. 

The daily new ads created between June 2015 to February 2016 were very low in comparison to March 2016 — only about 0.002059% to 0.006177% — and they were not created daily. In 2015, we also noticed that they were only 0.002059% of new ads on a single day in June, August, September and November, respectively; 0.002059 on two days in December; while none in July.

Based on these observations, we propose that **March or early spring is generally a good time for a buyer to search for a used car on eBay**.

In [47]:
# 'last_seen' — To get the distribution of date values as percentages
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()*100

2016-03-05     0.107075
2016-03-06     0.432419
2016-03-07     0.539494
2016-03-08     0.741290
2016-03-09     0.959559
2016-03-10     1.066634
2016-03-11     1.237542
2016-03-12     2.378305
2016-03-13     0.889548
2016-03-14     1.260193
2016-03-15     1.587596
2016-03-16     1.645252
2016-03-17     2.808665
2016-03-18     0.735112
2016-03-19     1.583477
2016-03-20     2.065316
2016-03-21     2.063257
2016-03-22     2.137386
2016-03-23     1.853225
2016-03-24     1.976773
2016-03-25     1.921176
2016-03-26     1.680257
2016-03-27     1.564945
2016-03-28     2.085907
2016-03-29     2.234165
2016-03-30     2.477144
2016-03-31     2.378305
2016-04-01     2.279466
2016-04-02     2.491558
2016-04-03     2.520385
2016-04-04     2.448316
2016-04-05    12.476320
2016-04-06    22.178980
2016-04-07    13.194959
Name: last_seen, dtype: float64

The distribution of `last_seen` percentage is rather consistent throughout the timeframe — it fluctuates within 0.107075% to 2.808665%. The percentage is a lot higher on the last three days, between 12.476320% to 22.178980%, which seems to be normal, as it is the date that the crawler saw the ad last online.

## Dealing with Incorrect Registration Year Data

Now, let's analyze the registration year data!

In [48]:
autos['registration_year'].describe()

count    48564.000000
mean      2004.755518
std         88.644797
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The descriptive statistics of `registration_year` comes to our surprise in two illogical aspects:
- The minimum registration year is 1000 — cars were not invented yet at that time!
- The maximum registration year is 9999 — year 9999 ridiculous!

We investigate the `registration_year` further to examine whether there are more unreasonable entries.

In [49]:
# To get the distribution of `registration_year` as percentages and sort the years in an ascending manner
autos['registration_year'].value_counts(normalize=True,dropna=False).sort_index()*100

1000    0.002059
1001    0.002059
1111    0.002059
1800    0.004118
1910    0.010296
1927    0.002059
1929    0.002059
1931    0.002059
1934    0.004118
1937    0.008237
1938    0.002059
1939    0.002059
1941    0.004118
1943    0.002059
1948    0.002059
1950    0.006177
1951    0.004118
1952    0.002059
1953    0.002059
1954    0.004118
1955    0.004118
1956    0.008237
1957    0.004118
1958    0.008237
1959    0.012355
1960    0.047360
1961    0.012355
1962    0.008237
1963    0.016473
1964    0.024710
          ...   
2000    6.496582
2001    5.427889
2002    5.119018
2003    5.557615
2004    5.565851
2005    6.045631
2006    5.497900
2007    4.688658
2008    4.560992
2009    4.293304
2010    3.271971
2011    3.341982
2012    2.697471
2013    1.653488
2014    1.365209
2015    0.807182
2016    2.512149
2017    2.866321
2018    0.967795
2019    0.004118
2800    0.002059
4100    0.002059
4500    0.002059
4800    0.002059
5000    0.008237
5911    0.002059
6200    0.002059
8888    0.0020

In addition to the year 1000, we notice that the unreasonable earliest registration years also include 1001, 1111, 1800 and 1910. Since the year 1886 is regarded as the birth year of the modern car (Ref: [modern car](https://en.wikipedia.org/wiki/Car)), any entries before the year 1910 in this dataset should be removed. 

Additionally, any entries after 2016 (the crawling year), i.e. between 2017 to 9999, should be removed as well.

In [50]:
# Remove the entries before the year 1910 and after the year 2016
autos = autos[autos['registration_year'].between(1910,2016)]

From the cell below, we can see that we have successfully removed the illogical years. Now, the minimum year is 1910, whereas the maximum year is 2016.

In [51]:
autos['registration_year'].describe()

count    46680.000000
mean      2002.910818
std          7.185168
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [52]:
# To get the distribution of `registration_year` as percentages and sort the years in a descending manner
autos['registration_year'].value_counts(normalize=True).sort_values(ascending=False).head(10)*100

2000    6.758783
2005    6.289632
1999    6.206084
2004    5.790488
2003    5.781919
2006    5.719794
2001    5.646958
2002    5.325621
1998    5.062125
2007    4.877892
Name: registration_year, dtype: float64

In [53]:
# To get the distribution of `registration_year` as percentages and sort the years in an ascending manner
autos['registration_year'].value_counts(normalize=True).sort_values().head(30)*100

1952    0.002142
1953    0.002142
1943    0.002142
1929    0.002142
1931    0.002142
1938    0.002142
1948    0.002142
1927    0.002142
1939    0.002142
1955    0.004284
1957    0.004284
1934    0.004284
1951    0.004284
1941    0.004284
1954    0.004284
1950    0.006427
1962    0.008569
1937    0.008569
1958    0.008569
1956    0.008569
1910    0.010711
1959    0.012853
1961    0.012853
1963    0.017138
1964    0.025707
1965    0.036418
1975    0.038560
1969    0.040703
1976    0.044987
1977    0.047129
Name: registration_year, dtype: float64

Based on the distribution, we can see that the top 10 `registration_year` are between 1998 to 2007 (4.88% to 6.76%). There was also a small percentage of sellers selling their antique cars on eBay site.

## Dealing with Incorrect Registration Month Data

Next, we will analyze the registration month data.

In [54]:
autos['registration_month'].describe()

count    46680.000000
mean         5.827078
std          3.670325
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [55]:
# To get the distribution of `registration_month` as percentages and sort the months in a descending manner
autos['registration_month'].value_counts(normalize=True).sort_values(ascending=False)*100

3     10.364182
6      8.823907
0      8.624679
4      8.341902
5      8.305484
7      7.973436
10     7.487147
12     6.988003
9      6.947301
11     6.917309
1      6.651671
8      6.469580
2      6.105398
Name: registration_month, dtype: float64

We are surprised by the fact that 8.62% of the registration was done in the month-*0*, which does not exist. Given that registration month is usually not so crucial compared to registration year in a market place, sellers might fill in month-*0* when they do not remember the registration month. We will just keep the entries for the month-*0* as they are, as registration month will not have a big impact on our analysis and we do not want to lose the associated data of the month-*0*.

Based on the data, **most of the cars were registered between spring and early summer (March to June)** — 8.31% to 10.36% — presumably preparing for summer holidays or spring/summer outdoor activities. Therefore, spring and early summer is the best time to change cars, especially March, as March is the peak time for ads creation as well.

To have an overview of the number of entries, we print it out:

In [56]:
n_before_removal = autos.shape[0]
print('The number of entries before removal of illogical registration months:   ', n_before_removal)

The number of entries before removal of illogical registration months:    46680


Now, let's examine whether there are any entries with illogical registration months. For example, it is illogical if an ad was created and crawled in March 2016, but the car was registered in August 2016.

The last ad creation date is 2016-04-07. In theory, the chance of someone buying and registering a new car and sell it within the same month is extremely low. Therefore, we create a combined Boolean filter for car registration after March 2016 by using `autos[(autos['registration_year'] == 2016) & (autos['registration_month'] > 3)]`, and remove the entries from autos by using `autos.drop()` method.

In [57]:
# Remove the entries of car registration after March 2016
autos = autos.drop(autos[(autos['registration_year'] == 2016) & (autos['registration_month'] > 3)].index)

autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
...                    ...                                                ...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

As we can see from the cell below, the number of entries have been reduced from 46680 to 45979 after removing 701 illogical entries.

In [58]:
n_after_removal = autos.shape[0]
print('The number of entries after removal of illogical registration months:   ', n_after_removal)
print('The number of deleted entries:    ', n_before_removal - n_after_removal)

The number of entries after removal of illogical registration months:    45979
The number of deleted entries:     701


While we check the entries for the registration year 2016, we only obtain the entries for registration month 0 to 3. This result further confirms the removal of the entries of the illogical registration months.

In [59]:
# Verify the deletion
autos[autos['registration_year'] == 2016]

date_crawled                                               name  \
59     2016-03-17 17:50:54               Mercedes_A_Klasse_W_168__A_140_gruen   
76     2016-03-22 14:52:57                         BMW_318i_neustes_Model_0Km   
101    2016-03-22 17:51:49    Schnaepchen_in_einem_Jahr_OLDTEIMER_KENNZEICHEN   
227    2016-04-04 17:54:21                               Als_Bastler_Fahrzeug   
343    2016-03-08 21:45:54                                    mitsubishi_colt   
374    2016-03-15 18:42:52                          Polo_9N_1_9_Tdi_Gti_Optik   
452    2016-03-08 10:52:14            Blauer_Golf_III_schoenes_Anfaenger_Auto   
479    2016-03-08 07:55:41                              Renault_R4_TL_Savanne   
696    2016-03-08 03:55:14                           Audi_a6_avant_2.5_tdi_v6   
1012   2016-03-05 21:49:36                                     Opel_Agila_1.0   
1124   2016-03-28 23:39:33                                       opel_corsa_A   
1133   2016-03-14 16:47:42                 Fiat_Panda_HP_100_Klimaautomatic!!   
1135   2016-03-19 12:45:25                                 Renault_twingo_1.3   
1306   2016-04-01 20:37:43           Peugeot_607_HDI_2.2_mit_gruener_Plakette   
1530   2016-03-25 15:50:38                                  Passat_35i_1_9_td   
1555   2016-03-17 01:00:21  Audi_Cabriolet____A5___NAVI_.XEN_._ALCANTARA.S...   
1562   2016-03-19 20:54:16  RENAULT_TWINGO___ERST_3_MONATE_ALT_UND_2000_KM...   
1632   2016-03-29 23:48:06                                 Fiat_Idea_Mini_Van   
1649   2016-03-25 13:43:29  Opel_Corsa_SPORT_GSi_16V_106_PS_mit_Ganzleder_...   
1866   2016-03-07 17:42:12  MICROCAR_M.GO_Dynamic_Mopedauto_Leichtkraftfah...   
1956   2016-03-07 15:46:32                          Volkswagen_Passat_1_9_tdi   
1988   2016-03-17 20:36:21                      Polo_9n3_1.6l_105PS_Automatik   
2278   2016-03-15 18:53:59                           Bmw_318_Coupe_M_Original   
2280   2016-03-31 17:51:37                                        Ford_fiesta   
2516   2016-03-15 09:57:39                                         Gutes_Auto   
2534   2016-03-07 09:57:07                                      Golf_1_Cabrio   
2583   2016-03-16 19:55:27          BMW_E46_Touring_320d_150ps_Leder_Tempomat   
2596   2016-03-10 18:49:15  Peugeot_206_75_Klima__8__fach_bereift_Scheckhe...   
2815   2016-03-31 10:53:50                           Volkswagen_Golf_4_unfall   
3121   2016-03-08 10:52:51             Ford_Focus_Zweitwagen_Anfaengerauto_VW   
...                    ...                                                ...   
46586  2016-03-21 16:54:39                         VW_Golf_VI_zu_verkaufen!!!   
46664  2016-03-12 07:36:59                      BMW_323_170ps_Leder_Automatik   
46782  2016-03-23 16:47:47  Audi_A4_Avant_S_line_S4_Optik_S_tronic_LED_Matrex   
46977  2016-03-09 14:50:31                                  Auto_zu_verkaufen   
46994  2016-04-03 15:50:56                                   opel_corsa_B_1.2   
46998  2016-03-11 08:37:34                                   MINI_One_Clubman   
47249  2016-03-12 23:57:17                                         Opel_corsa   
47270  2016-04-02 14:51:55                                    Golf_4_Gti_1.8t   
47414  2016-03-15 03:03:07                               Verkaufe_Honda_Civic   
47539  2016-03-14 18:25:17                         1_Hand_Civic_mit_Rest_Tuev   
47671  2016-03-11 13:37:40                                   Opel_corsa_b_1.0   
47832  2016-03-23 14:49:36  Opel_Vivaro_1.6_BiTurbo_2xSchiebetuer_KeyLess_...   
48254  2016-03-17 15:51:12                                  Ford_Focus_2_Tdci   
48275  2016-03-08 19:37:00                           FIAT_PUNTO_188_..._60_PS   
48437  2016-03-16 21:55:59                                 Golf_4_EDITION_1.4   
48559  2016-03-14 08:53:15                                  Audi_A3_Automatik   
48656  2016-03-28 01:54:01           Vectra_b_Bastler_Fahrzeug___auch_Tausch_   
48768  2016-04-03 18:54:34                   

## Exploring Price and Mileage by Brand

Now, we are going to analyze the price and mileage of cars by brand.

First, we explore the unique values in the `brand` column and sort them by using `value_counts()`:

In [60]:
print('Unique brand:')
print(autos['brand'].unique())
print('\n')
print('The number of unique brand:   ', len(autos['brand'].unique()))

Unique brand:
['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


The number of unique brand:    40


In [61]:
# Display the percentage of values counts of all car brands
autos['brand'].value_counts(normalize=True).sort_values(ascending=False)*100

volkswagen        21.066139
bmw               11.065921
opel              10.694012
mercedes_benz      9.687031
audi               8.699624
ford               6.972748
renault            4.708671
peugeot            2.973096
fiat               2.562039
seat               1.813872
skoda              1.650754
nissan             1.522434
mazda              1.513735
smart              1.409339
citroen            1.394115
toyota             1.272320
hyundai            1.000457
sonstige_autos     0.989582
volvo              0.919985
mini               0.880837
mitsubishi         0.822114
honda              0.782966
kia                0.715544
alfa_romeo         0.663346
porsche            0.617673
suzuki             0.591574
chevrolet          0.572000
chrysler           0.350160
dacia              0.263164
daihatsu           0.252289
jeep               0.228365
subaru             0.213141
land_rover         0.210966
saab               0.167468
jaguar             0.158768
daewoo             0

Out of the 40 unique brands, we decided to aggregate the top 10 brands for our analysis, as it represents a good proportion of the common car brands (1.81% to 21.07%). The brands that we have chosen are as below:

In [62]:
# Display the percentage of values counts of the top 10 car brands
autos['brand'].value_counts(normalize=True).sort_values(ascending=False).head(10)*100

volkswagen       21.066139
bmw              11.065921
opel             10.694012
mercedes_benz     9.687031
audi              8.699624
ford              6.972748
renault           4.708671
peugeot           2.973096
fiat              2.562039
seat              1.813872
Name: brand, dtype: float64

We use aggregation to understand the `brand` column. We start with grouping the the top 10 unique brands under the variable `top_10_brands`:

In [63]:
# Group the top 10 unique brands
top_10_brands = autos['brand'].value_counts(normalize=True).sort_values(ascending=False).head(10).index

print(top_10_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


Next, we loop over the brand names in `top_10_brands` to group the dataframe by unique brand names. By using the unique brand names, we compute the `brand_mean_price` and `brand_mean_mileage` for each of the unique brands, add them to the `mean_prices_dict` and `mean_mileage_dict` dictionaries, respectively.

In [64]:
# A dictionary to hold the aggegate data
mean_prices_dict = {}
mean_mileage_dict = {}
mean_prices_list = []

# Loop over the top 10 unique brands
for brand_name in top_10_brands:
    brand_groups = autos[autos['brand'] == brand_name]    # Group the rows of the unique brands
    
    # For mean price
    brand_mean_prices = int(brand_groups['price_$'].mean())    # Calculate the mean price of each of the unique brands and convert it to integer
    mean_prices_dict[brand_name] = brand_mean_prices    # Add to dictionary: key = brand_name; value = brand_mean_prices
    
    # For mean mileage
    brand_mean_mileage = int(brand_groups['odometer_km'].mean())    # Calculate the mean mileage of each of the unique brands and convert it to integer 
    mean_mileage_dict[brand_name] = brand_mean_mileage    # Add to dictionary: key = brand_name; value = brand_mean_mileage

print('Mean price dictionary:', '\n', mean_prices_dict)
print('\n')
print('Mean mileage dictionary:', '\n', mean_mileage_dict)


Mean price dictionary: 
 {'volkswagen': 5450, 'bmw': 8378, 'opel': 2995, 'mercedes_benz': 8688, 'audi': 9390, 'ford': 3795, 'renault': 2496, 'peugeot': 3109, 'fiat': 2833, 'seat': 4447}


Mean mileage dictionary: 
 {'volkswagen': 128526, 'bmw': 132466, 'opel': 129221, 'mercedes_benz': 130638, 'audi': 129041, 'ford': 124042, 'renault': 127886, 'peugeot': 126850, 'fiat': 116943, 'seat': 120803}


Now, we will investigate the correlation between the and the mean price and mean mileage by brand, if any. To do so, we will: 
- Construct two series objects.
- Combine the data from both series objects into a single dataframe and display the dataframe directly.

In [65]:
# Construct mean price series
mean_price_series = pd.Series(mean_prices_dict)

# Construct mean mileage series
mean_mileage_series = pd.Series(mean_mileage_dict)

print('** Mean price series **')
print(mean_price_series)
print('\n')

print('** Mean mileage series **')
print(mean_mileage_series)

** Mean price series **
volkswagen       5450
bmw              8378
opel             2995
mercedes_benz    8688
audi             9390
ford             3795
renault          2496
peugeot          3109
fiat             2833
seat             4447
dtype: int64


** Mean mileage series **
volkswagen       128526
bmw              132466
opel             129221
mercedes_benz    130638
audi             129041
ford             124042
renault          127886
peugeot          126850
fiat             116943
seat             120803
dtype: int64


In [66]:
# Construct a dataframe from `mean_price_series`
mean_price_mileage_df = pd.DataFrame(mean_price_series, columns=['mean_price_$'])

# Assign `mean_mileage_series` as a new column in this dataframe
mean_price_mileage_df['mean_mileage_km'] = pd.DataFrame(mean_mileage_series)

# Sort the mean price in a descending manner
mean_price_mileage_df.sort_values(['mean_price_$'], ascending=False)

mean_price_$  mean_mileage_km
audi                   9390           129041
mercedes_benz          8688           130638
bmw                    8378           132466
volkswagen             5450           128526
seat                   4447           120803
ford                   3795           124042
peugeot                3109           126850
opel                   2995           129221
fiat                   2833           116943
renault                2496           127886

In [67]:
# Sort the mean mileage in a descending manner
mean_price_mileage_df.sort_values(['mean_mileage_km'], ascending=False)

mean_price_$  mean_mileage_km
bmw                    8378           132466
mercedes_benz          8688           130638
opel                   2995           129221
audi                   9390           129041
volkswagen             5450           128526
renault                2496           127886
peugeot                3109           126850
ford                   3795           124042
seat                   4447           120803
fiat                   2833           116943

In [68]:
max_mileage = mean_price_mileage_df['mean_mileage_km'].max()
min_mileage = mean_price_mileage_df['mean_mileage_km'].min()

difference_mileage_percent = round((max_mileage - min_mileage) * 100 / max_mileage, 1) 

print('The difference between the maximum and minimum mileage:   ', difference_mileage_percent, '%')

The difference between the maximum and minimum mileage:    11.7 %


We notice that there are three distinct price gaps between the 10 top brands:
- More expensive (> 8000): Audi, Mercedes-Benz, BMW
- Moderate expensive (4000 - 7999): Volkswagen, Seat
- Less expensive (< 4000): Ford, Peugeot, Opel, Fiat, Renault

The difference in mileage between the maximum and minimum prices among the top 10 brands is not big, only 11.7%.

Based on the comparison result, we **do not observe any correlation between the mean price and the mean mileage by brand**. In fact, the mean price of the brands is possibly determined by more complicated combined factors, such as credibility of the brand, marketing strategy, popularity in certain countries/regions, car specifications, etc.

## Conclusion

In this project, we cleaned the used cars dataset from eBay Kleinanzeigen, by removing the non-value-added columns, outliers in the price column and the illogical registration year and month data. We further analyzed the car listings and discovered :
- The highest count frequency for odometer is 150,000 km
- The distribution of `date_crawled` is fairly consistent throughout the whole crawling period
- The peak time for `ads_created` is March 2016 and beginning of April 2016 
- The distribution of `last_seen` percentage is rather consistent throughout the timeframe
- No correlation between mean price and mean mileage by brand